frontend/src/Onboarding/TransferCoursesScreen.tsx

```diff
@@ -35,253 +30,159 @@ const TitleText = styled.div`
-            listReq.subject === req[0].subject
-        )
-      )
-      .map((r: IRequiredCourse[]) => this.renderCourse(r));
-  }
+  const CourseSet: React.FC<{ courses: IRequiredCourse[] }> = ({ courses }) => {
+    const filtered = selectedRequirements.filter(
+      r => !courses.some(c => courseEq(r, c))
+    );
+    return (
+      <ControlledSelectableCourse
+        checked={filtered.length !== selectedRequirements.length}
+        courseText={courses
+          .map(({ subject, classId }) => subject + classId)
+          .join(" and ")}
+        onChange={e =>
+          setSelectedRequirements(
+            e.target.checked ? [...selectedRequirements, ...courses] : filtered
+          )
+        }
+      />
+    );
+  };
 
   /**
    * renders an entire requirement section if it has  classes specified
    * with the title of the section
    * Only classes that have already been selected as a completed course will be rendered
    * @param requirementGroup - the requirement group to be rendered
    */
-  renderSection(requirementGroup: string) {
-    const reqs = this.props.major.requirementGroupMap[requirementGroup];
+  const RequirementSection: React.FC<{ requirementGroup: string }> = ({
+    requirementGroup,
+  }) => {
+    const reqs = major.requirementGroupMap[requirementGroup];
     if (!reqs || reqs.type === "RANGE") {
       return null;
     }
-    const renderedReqs = this.parseCourseRequirements(
-      flatten(reqs.requirements)
+    const filteredReqs = flatten(reqs.requirements).filter(req =>
+      completedRequirements.some(listReq => courseEq(listReq, req[0]))
     );
-    const returnDiv = (
-      <div key={requirementGroup}>
-        <TitleText>{requirementGroup}</TitleText>
-        {renderedReqs}
-      </div>
-    );
-    return renderedReqs.length > 0 ? returnDiv : null;
-  }
 
-  // renders the "Other Course" section with a button to display the add coursese modal and
-  // displays all already added courses under the button.
-  renderOtherCourseSection() {
+    if (filteredReqs.length === 0) return null;
     return (
-      <div key="other courses">
-        <TitleText>Other Courses</TitleText>
-        <AddBlock onClick={this.showModal.bind(this)} />
-        {this.state.otherCourses.map(this.renderCourse.bind(this))}
+      <div key={requirementGroup}>
+        <TitleText>{requirementGroup}</TitleText>
+        {filteredReqs.map(c => (
+          <CourseSet courses={c} key={coursesToString(c)} />
+        ))}
       </div>
     );
-  }
-
-  /**
-   * Renders one course/courseset (if it contains labs/recitiations, and seperated)
-   * @param courses - Course pairings to be rendered
-   */
-  renderCourse(courses: IRequiredCourse[]) {
-    let allCourse = courses.map(course => course.subject + course.classId);
-    return (
-      <SelectableCourse
-        courseText={allCourse.join(" and ")}
-        onChange={e => courses.forEach(course => this.onChecked(e, course))}
```

Comment:
bug # 3 part 2: really this should be _one_ state update for all the courses in the set, not several updates (one per course). the later setState calls will override the beginning ones

Reply:
sweet!